Faster listing of stake pools.

jonathanknowles opened the pull request:
## Related Issues

#2082 (_Listing stake pools time grows when there are wallets syncing in background_.)

https://jira.iohk.io/browse/ADP-383 (_Use more efficient DB queries when listing pools_.)

## Overview

This PR:

* changes the implementation of `ListStakePools` to execute a small number of handwritten SQL queries with native joins, rather than issuing multiple queries and joining together the results in memory.

* adds a new operation [`listPoolLifeCycleData`](https://github.com/input-output-hk/cardano-wallet/blob/d00c608e47d2d189a282a02946cc8e46b843375b/lib/core/src/Cardano/Pool/DB.hs#L180). The [SQLite implementation](https://github.com/input-output-hk/cardano-wallet/blob/d00c608e47d2d189a282a02946cc8e46b843375b/lib/core/src/Cardano/Pool/DB/Sqlite.hs#L384) of this operation uses a [single hand-optimized query](https://github.com/input-output-hk/cardano-wallet/blob/d00c608e47d2d189a282a02946cc8e46b843375b/lib/core/src/Cardano/Pool/DB/Sqlite.hs#L595) to return lifecycle data for all pools.

## Benefits

### :star: **Faster execution time**

  The time taken to execute `ListStakePools` is greatly reduced, even when there are multiple wallets syncing in the background.
  <details><summary>Click to view data</summary><br>

  | No. of <br>Wallets<br>Syncing | Time Required<br>(seconds)<br>`master` | Time Required<br>(seconds)<br>`more-efficient-list-pools` |
  | ---: | ---: | ---: |
  | 0 | 8 | 3 |
  | 5 | 30 | 7 |
  | 10 | 60 | 8 |
  | 15 | 72 | 9 |
  | 20 | 95 | 10 |
  </details>

### :star: **Fewer database queries**

  The number of `select` queries required for a single call to `ListStakePools` is greatly reduced.
  <details><summary>Click to view data</summary><br>
  
  | Branch | Network | No. of Pools | No. of Queries |
  | :--- | :--- | ---: | ---: | 
  | `master` | `mainnet` |  1,124 |  12,365 |
  | `more-efficient-list-pools`| `mainnet` | 1,124  | 4 |
  </details>

## Testing

### :heavy_check_mark: Properties

This PR adds a [property test](https://github.com/input-output-hk/cardano-wallet/blob/d00c608e47d2d189a282a02946cc8e46b843375b/lib/core/test/unit/Cardano/Pool/DB/Properties.hs#L987) for `listPoolLifeCycleData`, consisting of the following steps:
1. Add an arbitrary sequence of pool registration and retirement certificates to the database, for multiple pools.
2. Call `listPoolLifeCycleData` once to fetch lifecycle data for all active pools.
3. Call `readPoolLifeCycleStatus` multiple times, once for each known pool, and coalesce the results. 
4. Test that the results returned in steps 2 and 3 are identical.

### :heavy_check_mark: Live Data

Run the following command to get a list of all stake pools sorted by ID:
```sh
cardano-wallet stake-pool list --stake 1000 | jq 'sort_by(.id)' > stake-pool-list
```
The results should be identical for both `master` and this branch.

## Notes

### System used for Benchmarking

The system used for benchmarking (4 cores):

```
vendor_id       : GenuineIntel
cpu family      : 6
model           : 85
model name      : Intel(R) Xeon(R) CPU
stepping        : 7
microcode       : 0x1
cpu MHz         : 2800.176
cache size      : 33792 KB
```

jonathanknowles committed:
Add `active_pool_registrations` pool database view.

jonathanknowles committed:
Add `active_pool_lifecycle_data` pool database view.

jonathanknowles committed:
Add `active_pool_owners` pool database view.

jonathanknowles committed:
Pass `currentEpoch` through to `readPoolDbData`.

This makes it possible to filter out retired pools at the SQL query level.

jonathanknowles committed:
Remove in-memory filtering of retired pools.

This is no longer necessary, as retired pools are now filtered out at
the SQL query level.

jonathanknowles committed:
Add operation `listPoolLifeCycleData` to pool database layer.

Both SQLite and model implementations are provided.

jonathanknowles committed:
Use `listPoolLifeCycleData` within `readPoolDbData`.

This replaces a sequence of n database operations with a single database
operation that performs just a single query.

jonathanknowles committed:
Use a single aggregate query for `readTotalProduction`.

jonathanknowles committed:
Add `MultiPoolCertificateSequence` type to test suite.

This type will be used by later commits, to generate sequences of
registration and retirement certificates affecting multiple pools.

jonathanknowles committed:
Simplify `prop_listRetiredPools_multiplePools_multipleCerts`.

Use `MultiPoolCertificateSequence` to simplify the
`prop_listRetiredPools_multiplePools_multipleCerts` property.

jonathanknowles committed:
Add `ListSerializationOrder` to `MultiPoolCertificateSequence`.

This makes it possible to provide the following function, which produces
a single, flattened list of pool certificates affecting multiple pools.

```
getMultiPoolCertificateSequence
    :: MultiPoolCertificateSequence -> [PoolCertificate]
```

jonathanknowles committed:
Further simplify `prop_listRetiredPools_multiplePools_multipleCerts`.

jonathanknowles committed:
Introduce function `putPoolCertificate` to reduce code duplication.

jonathanknowles committed:
Introduce `testCertificatePublicationTimes` to reduce code duplication.

jonathanknowles committed:
Allow `PoolLifeCycleStatus` values to be included in sets.

jonathanknowles committed:
Do not generate registration certificates with empty owners lists.

Every valid registration certificate must have at least one owner.

Ideally, this should be encoded in the type system, but we make do with
not generating invalid data for now.

jonathanknowles committed:
Rename helper function to `firstCertificateIsNotRetirement`.

This more accurately reflects the function's purpose.

We don't actually mind if the list of certificates is empty.

jonathanknowles committed:
Add `prop_listPoolLifeCycleData_multiplePools_multipleCerts`.

hasufell commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -208,14 +211,20 @@ newDBLayer trace fp timeInterpreter = do
-
-            pure $ Map.map Quantity $ foldl' toMap Map.empty production
+        readTotalProduction =
+            Map.fromList . rights . fmap parseRow <$> rawSql query []
```
This looks like we hide parse errors?

jonathanknowles replied:
> This looks like we hide parse errors?

For this particular function, I think it would be appropriate to fail fast if there's even a single parse error. It's a very simple `group by` query that should not have any failures at all.

jonathanknowles replied:
Fixed in 6fbc807c47cfea70f3f7c7e4fa628048cc703ac6.

hasufell commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -372,6 +381,55 @@ newDBLayer trace fp timeInterpreter = do
             rights . fmap safeCast <$> rawSql query parameters
 
+        listPoolLifeCycleData epochNo =
+            rights . fmap parseRow <$> rawSql query parameters
```
Same here, looks like we hide parse errors.

jonathanknowles replied:
> looks like we hide parse errors.

It does indeed!

I previously added a patch that would cause this function to fail fast in the face of a parse error. However, that had the disadvantage that a single row failing to parse would cause the entire operation to fail.

After giving this some thought, I think the following method would be more appropriate, especially from a user experience perspective:

- for rows that fail to parse: create a log entry for each error.
- for rows that parse successfully: return these as part of the result.

jonathanknowles replied:
Fixed in 79f505c41ab68c2bc6d134b43071163add14a3be.

jonathanknowles committed:
Log parse failures in pool DB operations.

Log parse failures in the following pool DB operations:
  - `readTotalProduction`.
  - `listPoolLifeCycleData`.
  - `listRetiredPools`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2111#discussion_r484482133

jonathanknowles committed:
Add `MsgParseFailure` constructor to `PoolDbLog`.

jonathanknowles committed:
Define `knownPools` in terms of `listRegisteredPools`.

There's no point in fetching the entire set of all pool-related data
when all we really need is the set of valid pool IDs.

hasufell commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -208,14 +211,24 @@ newDBLayer trace fp timeInterpreter = do
+            assertSuccess = flip either id $ \e -> error $ mconcat
+                [ "readTotalProduction: unexpected parse failure: "
+                , show e
+                ]
```
Since we're in IO it's best to throw it as IO errors. `error` triggers on evaluation and can leak through exception handlers. Probably not a big concern here, but better safe than sorry.

```suggestion
        readTotalProduction = do
            rows <- rawSql query []
            parsed <- forM rows $ \row -> case parseRow row of
                Right x -> pure x
                Left e  -> liftIO . throwIO . userError
                    . ("readTotalProduction: unexpected parse failure: " ++)
                    . T.unpack $ e
            pure $ Map.fromList parsed
          where
```

jonathanknowles replied:
Thanks @hasufell, I agree it's better to record parse failures eagerly, as soon as they happen, rather than relying on the lazy evaluation of `error`.

In the end I changed this to use the logging mechanism instead, as I think this will make any errors easier to analyze.

hasufell replied:
Logging doesn't short-circuit and doesn't crash the handler, though. From what I remember, we don't expect parse failures here ever?

jonathanknowles committed:
Add function `runRawQuery` to reduce code duplication.

jonathanknowles committed:
Always recreate views when starting the pool worker.

This makes it possible to change the definitions of views as required.

rvl reviewed (approved):
This is a good improvement.

Automated performance testing is always good... a simple count of queries should be enough in most cases. Perhaps it's possible to test that a call to `listPoolLifeCycleData` never causes more than (say) 5 SQL queries.

Let's have the squabble about boring SQL on slack.
